equational-reasoning.md

```diff
@@ -0,0 +1,29 @@
+Equational reasoning
+====================
+
+**Tags**: *substitution model*, *referential transparency*, *purity*, *side effect*
+
+Concerning programming, we can say that a function definition limits the scope where different expressions are evaluated in order to produce a result. This result is produced by the transformation of its input parameters (if they exist) along with any other internal variables and defines the return type of that function.
+
+For developers it seems very convenient to be able to `reason` about a function (i.e. understand what the function does ) just by checking the input and the output of the function [[1]](#comments). This is to say that a function's return value represents the transformation of its internal state by potentially using other arguments as its parameters.
```

Comment:
_ a function's return value represents the transformation of its internal state by potentially using other arguments as its parameters_ - what is the return value of the function `x => x + 1`? What is its internal state?

Reply:
I would say that initially its state is x and then it changes to x + 1

Reply:
Dimitrios is touching an important concept from which the State Monad comes from, the fact that you can see a function as a state transition. Maybe it is slightly confusing when you say "internal state", because here "internal" is being a little bit misleading, one could think that you refer to an inner, mutable, persistent state.